rust_snuba/rust_arroyo/src/backends/kafka/mod.rs

```diff
@@ -301,41 +281,30 @@ impl<C: AssignmentCallbacks> ArroyoConsumer<KafkaPayload, C> for KafkaConsumer<C
         Ok(())
     }
 
-    fn stage_offsets(&mut self, offsets: HashMap<Partition, u64>) -> Result<(), ConsumerError> {
-        for (partition, offset) in offsets {
-            self.staged_offsets.insert(partition, offset);
-        }
-        Ok(())
-    }
-
-    fn commit_offsets(&mut self) -> Result<HashMap<Partition, u64>, ConsumerError> {
+    fn commit_offsets(&mut self, offsets: HashMap<Partition, u64>) -> Result<(), ConsumerError> {
         self.state.assert_consuming_state()?;
 
-        let mut partitions = TopicPartitionList::with_capacity(self.staged_offsets.len());
-        for (partition, offset) in &self.staged_offsets {
+        let mut partitions = TopicPartitionList::with_capacity(offsets.len());
+        for (partition, offset) in &offsets {
             partitions.add_partition_offset(
                 partition.topic.as_str(),
                 partition.index as i32,
                 Offset::from_raw(*offset as i64),
             )?;
         }
 
-        let mut consumer = self.consumer.lock().unwrap();
-        consumer
-            .as_mut()
+        self.consumer
+            .as_ref()
             .unwrap()
             .commit(&partitions, CommitMode::Sync)
             .unwrap();
 
-        // Clear staged offsets
-        let prev_offsets = mem::take(&mut self.staged_offsets);
-
-        Ok(prev_offsets)
+        Ok(())
     }
 
     fn close(&mut self) {
         self.state = KafkaConsumerState::Closed;
-        *self.consumer.lock().unwrap() = None;
+        self.consumer = None;
```

Comment:
With the circular references cleaned up, we might also be able to get rid of that `Option` and `unwrap` everywhere by constructing the `BaseConsumer` in the constructor, and having a proper `fn close(self)` as well.

Reply:
Unfortunately it's not that simple. We can't construct the `BaseConsumer` in `new` because we don't have the callbacks yet, we only get those during `subscribe` (via `StreamProcessor::subscribe`). The construction process currently looks like this:
1. Create a `KafkaConsumer` with no `BaseConsumer` inside
2. Create a `StreamProcessor` with the `KafkaConsumer` from (1)
3. Call `subscribe` on the `StreamProcessor`, which instantiates the callbacks and passes them to `KafkaConsumer::subscribe`, which uses them to create the `BaseConsumer`.

The fundamental problem here is that the `BaseConsumer` needs access to the callbacks, but the callbacks depend on the `StreamProcessor` (because they refer to its strategy).

Reply:
I believe the fundamental problem with that is the lifecycle difference between `new`/`subscribe` (and also `close`/`drop`). So we would have to find a way to create and subscribe in one go.